Fix build failures and expand test coverage

eleanorLYJ opened the pull request:
# Fix build issues and enhance build system tests
Improve our build system testing and fix various build issues discovered in the process

### CI Improvements
- Add compiler matrix to test with both gcc-latest and clang-latest
- Set warnings as errors to catch potential issues early
- Add diverse build configurations testing

### Build Fixes
- Fix linker error when using Clang by adding missing `$(LDFLAGS)`
- Resolve multiple compiler warning, including:
    - Enabling JIT with specific features disabled, such as Zba, Zbb, Zbc, Zbs, EXT_M, or Zicsr
    - Disabling Zicsr 
    - Disabling MOP_FUSIO
- Resolve multiple Clang-specific build errors, including:
    - Enabling JIT when EXT_M or EXT_C disabled
    - Disabling BLOCK_CHAINING
    - Disabling MOP FUSION
    - Disabling Zicsr

## Testing Done
- Test builds with various feature flags disabled
- Confirm CI pipeline runs successfully with all new test cases

eleanorLYJ committed:
CI: Add compiler matrix for gcc and clang testing

Add compiler matrix strategy to test builds with both gcc and clang
compilers. This ensures the codebase compiles and passes tests with
different compilers.

eleanorLYJ committed:
Fix compiler error when enabling JIT

When using make 'ENABLE_JIT=1', the following error is observed:

src/jit.c:1760:17: error: ‘do_sret’ defined but not used [-Werror=unused-function]
 1760 |     static void do_##inst(struct jit_state *state UNUSED, riscv_t *rv UNUSED, \
      |                 ^~~

Add "SYSTEM" to the EXT_LIST, which cause SERT to be add SKIP_LIST
during template generation. The prevents the generatoin of unused JIT
code for the SERT instruction.

Fixes: fc0d878 ("Preliminary support for trap handling during block emulation")

eleanorLYJ committed:
Fix compiler error with JIT when Zba is disabled

When using 'make ENABLE_Zba=0 ENABLE_JIT=1', the following errors are
observed:

src/jit.c:1760:17: error: ‘do_sh3add’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_sh2add’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_sh1add’ defined but not used [-Werror=unused-function]

Include 'Zba' in EXT_LIST to ensure instructions like 'sh3add',
'sh2add', and 'sh1add' are properly handled, resolving unused function
errors during JIT builds.

eleanorLYJ committed:
Fix compiler errors with JIT when Zbb is disabled

When using 'make ENABLE_Zbb=0 ENABLE_JIT=1', the following errors are
observed:

src/jit.c:1760:17: error: ‘do_rev8’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_orcb’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_rori’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_ror’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_rol’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_zexth’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_sexth’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_sextb’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_minu’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_maxu’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_min’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_max’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_cpop’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_ctz’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_clz’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_xnor’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_orn’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_andn’ defined but not used [-Werror=unused-function]

Include 'Zbb' in EXT_LIST to ensure instructions like 'rev8',
'orcb', and 'rori' are properly handled, resolving unused function
errors during JIT builds.

eleanorLYJ committed:
Fix compiler errors with JIT when Zbc is disabled

When using 'make ENABLE_Zbc=0 ENABLE_JIT=1', the following errors are
observed:

src/jit.c:1760:17: error: ‘do_clmulr’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_clmulh’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_clmul’ defined but not used [-Werror=unused-function]

Include 'Zbc' in EXT_LIST to ensure instructions like 'clmulr',
'clmulh', and 'clmul' are properly handled, resolving unused function
errors during JIT builds.

eleanorLYJ committed:
Fix compiler error with when Zicsr is disabled

When using 'make ENABLE_Zicsr=0', the following error is observed:

src/emulate.c:561:10: error: ‘rv_insn_csrrw’ undeclared (first use in this function); did you mean ‘rv_insn_csrai’?
  561 |     case rv_insn_csrrw:
      |          ^~~~~~~~~~~~~
      |          rv_insn_csrai

The error occurs because CSR instruction enums are only defined when
the Zicsr extension is enabled. By adding the feature guard, we ensure
the switch cases are only compiled when the required instruction
definitions are available.

eleanorLYJ committed:
Fix compiler error with JIT when Zicsr is disabled

When using 'make ENABLE_Zicsr=0 ENABLE_JIT=1', the following error is
observed:

src/jit.c:1760:17: error: ‘do_csrrwi’ defined but not used [-Werror=unused-function]
 1760 |     static void do_##inst(struct jit_state *state UNUSED, riscv_t *rv UNUSED, \
      |                 ^~~
src/rv32_jit.c:363:1: note: in expansion of macro ‘GEN’
  363 | GEN(csrrwi, {assert(NULL);
      | ^~~

This ensures csrrwi instructions are properly excluded from JIT code
generation when the Zicsr extension is disabled.

eleanorLYJ committed:
Fix compiler error with JIT when EXT_M is disabled

When using 'make ENABLE_EXT_M=0 ENABLE_JIT=1', the following error is
observed:

src/jit.c:1693:13: error: ‘ra_load2_sext’ defined but not used [-Werror=unused-function]
 1693 | static void ra_load2_sext(struct jit_state *state,
      |             ^~~~~~~~~~~~~

Fix this by guarding the ra_load2_sext function definition with
RV32_HAS(EXT_M) macro. This ensures the function is only defined when
the M extension is enabled.

eleanorLYJ committed:
Fix compiler error when MOP_FUSION is disabled

When using 'make ENABLE_MOP_FUSION=0', the following error is observed:

src/emulate.c:714:13: error: ‘match_pattern’ defined but not used [-Werror=unused-function]
  714 | static void match_pattern(riscv_t *rv, block_t *block)
      |             ^~~~~~~~~~~~~

Fix this by guarding the match_pattern function definition with
RV32_HAS(MOP_FUSION) macro. This ensures the function is only defined
when the MOP_FUSION is enabled.

eleanorLYJ committed:
CI: Add test case for disabling Zba with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zba extension is disabled. This ensures better coverage and prevents
regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbb with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zbb extension is disabled. This ensures better coverage and prevents
regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbc with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zbc extension is disabled. This ensures better coverage and prevents
regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbs with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zbs extension is disabled. This ensures better coverage and prevents
regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zicsr

Add a new test case to the CI pipeline when the Zicsr extension is
disabled. This ensures better coverage and prevents regressions for
this specific configuration.

eleanorLYJ committed:
Fix compiler errors with JIT when Zbs is disabled

When using 'make ENABLE_Zbs=0 ENABLE_JIT=1', the following errors are
observed:

src/jit.c:1760:17: error: ‘do_bseti’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_bset’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_binvi’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_binv’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_bexti’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_bext’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_bclri’ defined but not used [-Werror=unused-function]
src/jit.c:1760:17: error: ‘do_bclr’ defined but not used [-Werror=unused-function
]
Include "Zbs" in EXT_LIST to ensure instructions like `bseti`, `bset`,
and `bclr`  are properly handled, resolving unused function errors
during JIT builds.

eleanorLYJ committed:
CI: Add test case for disabling EXT_M with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the EXT_M extension is disabled. This ensures better coverage and
prevents regressions for this specific configuration.

eleanorLYJ committed:
Treat warnings as errors by adding -Werror

Add the -Werror flag to the CFLAGS to treat all warnings as errors
during the build process. This ensures that warnings are blocked by CI,
enforcing cleaner code and preventing builds from succeeding with any
warnings

eleanorLYJ committed:
CI: Add test case for disabling MOP FUSION

Add a new test case to the CI pipeline when the MOP FUSION is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
CI: Add test case for disabling BLOCK CHAINGING

Add a new test case to the CI pipeline when the BLOCK_CHAINING is
disabled. This ensures better coverage and prevents regressions for
this specific configuration.

eleanorLYJ committed:
Fix Clang error when using 'make tests'

When using 'make tests CC=clang', the following error is observed:
build/cache/test-cache.o: file not recognized: file format not recognized
clang: error: linker command failed with exit code 1 (use -v to see invocation)

Adding $(LDFLAGS) resolves the issue, ensuring successful linking.

eleanorLYJ committed:
CI: Add test case for disabling Zba

Add a new test case to the CI pipeline when the Zba is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbb

Add a new test case to the CI pipeline when the Zbb is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbc

Add a new test case to the CI pipeline when the Zbc is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zbs

Add a new test case to the CI pipeline when the Zbs is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
Fix Clang error when disabling Zicsr

When using 'make ENABLE_Zicsr=0 CC=clang', the following error is
observed:

src/emulate.c:83:20: error: unused function 'update_time' [-Werror,-Wunused-function]
   83 | static inline void update_time(riscv_t *rv)
      |                    ^~~~~~~~~~~
1 error generated

Fix this by guarding the update_time function definition with
RV32_HAS(Zicsr) macro. This ensures the function is only defined
when the Zicsr is enabled.

eleanorLYJ committed:
Fix Clang error when disabling MOP FUSION

When using 'make ENABLE_MOP_FUSION=0 CC=clang', the following error is
observed:
src/emulate.c:695:20: error: unused function 'remove_next_nth_ir' [-Werror,-Wunused-function]
  695 | static inline void remove_next_nth_ir(const riscv_t *rv,
	  |                    ^~~~~~~~~~~~~~~~~~
1 error generated.

Fix by gurading the 'remove_next_nth_ir' function and the
'COMBINE_MEM_OPS' macro with the 'RV32_HAS(MOP_FUSION)' macro. This
ensures both are only included when MOP FUSION is enabled, preventing the
Clang warning and maintaining proper functionality.

eleanorLYJ committed:
Fix Clang error when disabling BLOCK_CHAINING

When using 'make ENABLE_MBLOCK_CHAINING=0 CC=clang', the following
errors are observed:
src/emulate.c:555:19: error: unused function 'insn_is_unconditional_branch' [-Werror,-Wunused-function]
  555 | FORCE_INLINE bool insn_is_unconditional_branch(uint8_t opcode)
      |                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/emulate.c:582:19: error: unused function 'insn_is_direct_branch' [-Werror,-Wunused-function]
  582 | FORCE_INLINE bool insn_is_direct_branch(uint8_t opcode)
      |                   ^~~~~~~~~~~~~~~~~~~~~

Fix this by guarding the 'insn_is_unconditional_branch' and
'insn_is_direct_branch' function definition with
RV32_HAS(BLOCK_CHAINING) macro. This ensures the functions are only
defined when the BLOCK_CHAINING is enabled.

eleanorLYJ committed:
Fix Clang error with JIT when EXT_C is disabled

When using 'make make ENABLE_EXT_C=0 ENABLE_JIT=1 CC=clang', the
following errors are observed:

rc/t2c.c:77:1: error: unused function 't2c_gen_ra_addr' [-Werror,-Wunused-function]
   77 | T2C_LLVM_GEN_ADDR(ra, X, rv_reg_ra);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/t2c.c:64:31: note: expanded from macro 'T2C_LLVM_GEN_ADDR'
   64 |     FORCE_INLINE LLVMValueRef t2c_gen_##reg##_addr(                           \
      |                               ^~~~~~~~~~~~~~~~~~~~
<scratch space>:75:1: note: expanded from here
   75 | t2c_gen_ra_addr
      | ^~~~~~~~~~~~~~~
src/t2c.c:78:1: error: unused function 't2c_gen_sp_addr' [-Werror,-Wunused-function]
   78 | T2C_LLVM_GEN_ADDR(sp, X, rv_reg_sp);
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/t2c.c:64:31: note: expanded from macro 'T2C_LLVM_GEN_ADDR'
   64 |     FORCE_INLINE LLVMValueRef t2c_gen_##reg##_addr(                           \
      |                               ^~~~~~~~~~~~~~~~~~~~

Fix this by guarding the 't2c_gen_ra_addr' and 't2c_gen_sp_addr'
function definition with RV32_HAS(EXT_C) macro. This ensures the
functions are only defined when the EXT_C is enabled.

eleanorLYJ committed:
Fix Clang error with JIT when EXT_M is disabled

When using 'make ENABLE_EXT_M=0 ENABLE_JIT=1 CC=clang', the following
error is observed:

src/jit.c:728:20: error: unused function 'emit_alu64_imm8' [-Werror,-Wunused-function]
  728 | static inline void emit_alu64_imm8(struct jit_state *state,
      |                    ^~~~~~~~~~~~~~~

Fix this by guarding the 'emit_alu64_imm8' function definition with
RV32_HAS(EXT_M) macro. This ensures the function is only defined
when the EXT_M is enabled.

eleanorLYJ committed:
Fix undefined behavior caused by mismatched parameter types

When using 'make ENABLE_UBSAN=1 check CC=clang', the following error is
observed:

src/emulate.c:1110:13: runtime error: call to function do_fuse1 through pointer to incorrect function type 'bool (*)(struct riscv_internal *, const struct rv_insn *, unsigned long, unsigned int)'
/home/eleanor/code/rv32emu/src/emulate.c:415: note: do_fuse1 defined here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior src/emulate.c:1110:13

After fixing the first error with 'do_fuse1', similar errors were
observed for other functions like 'do_fuse2', 'do_fuse3', 'do_fuse4',
and 'do_jal'. The root cause was type mismatches in parameter
declarations, where 'rv_insn_t *' was used instead of the expected
'const rv_insn_t *'. Since 'do_jal' was generated by the 'RVOP'
macro in 'rv32emu_template.c', the macro was also corrected.

These changes resolve the UBSAN errors and align all function pointers
and implementations.

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: 99c52b9f4e89afb58a185ac9309a96cfee3ac17f | Previous: 3f37f514e33891e2d35bc614fb8d33035f2a4e24 | Ratio |
|-|-|-|-|
| `Dhrystone` | `1327` Average DMIPS over 10 runs | `1317` Average DMIPS over 10 runs | `0.99` |
| `Coremark` | `954.362` Average iterations/sec over 10 runs | `956.978` Average iterations/sec over 10 runs | `1.00` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

visitorckw commented on .github/workflows/main.yml:
```diff
@@ -49,43 +53,77 @@ jobs:
+            make ENABLE_JIT=1 clean && make ENABLE_Zba=0 ENABLE_JIT=1 check -j$(nproc)
+            make ENABLE_JIT=1 clean && make ENABLE_Zbb=0 ENABLE_JIT=1 check -j$(nproc)
+            make ENABLE_JIT=1 clean && make ENABLE_Zbc=0 ENABLE_JIT=1 check -j$(nproc)
+            make ENABLE_JIT=1 clean && make ENABLE_Zbs=0 ENABLE_JIT=1 check -j$(nproc)
```
How about repeating all diverse configuration tests with JIT enabled?

visitorckw commented on .github/workflows/main.yml:
```diff
@@ -49,43 +53,77 @@ jobs:
+            make distclean && make ENABLE_Zba=0 check -j$(nproc)
+            make distclean && make ENABLE_Zbb=0 check -j$(nproc)
+            make distclean && make ENABLE_Zbc=0 check -j$(nproc)
+            make distclean && make ENABLE_Zbs=0 check -j$(nproc)
```
Maybe also add:
`make distclean && make ENABLE_Zifencei=0 check -j$(nproc)`

visitorckw commented:
Hmmm, it seems CI fails due to insufficient memory. Any ideas on how we can reduce the memory usage?

henrybear327 commented on Makefile:
```diff
@@ -7,7 +7,7 @@ BIN := $(OUT)/rv32emu
 CONFIG_FILE := $(OUT)/.config
 -include $(CONFIG_FILE)
 
-CFLAGS = -std=gnu99 -O2 -Wall -Wextra
```
Since we are modifying this, do we also want to add `-Wshadow` to catch more problems :) ?

visitorckw replied:
I think this depends on whether our coding style aims to forbid local variables from shadowing other variables?

jserv commented on .github/workflows/main.yml:
```diff
@@ -36,6 +36,10 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        compiler: [gcc-latest, clang-latest]
```
Any reason why ordinary `gcc` and `clang` fail to work?

eleanorLYJ replied:
Are you asking if it's better to avoid using the latest gcc/clang and instead use older, more stable version?

jserv replied:
> Are you asking if it's better to avoid using the latest gcc/clang and instead use older, more stable version?

Yes, using the clang/gcc compilers from official Linux distribution packages makes it easier to track and report bugs, as we can identify build failures specific to the distribution versions we use in our CI pipeline.

eleanorLYJ replied:
Do you have a preference for a specific version of gcc or clang?

jserv replied:
> Do you have a preference for a specific version of gcc or clang?

Use the versions of gcc/clang provided by selected Linux distributions.

eleanorLYJ committed:
Fix unused variable warning when JIT and SYSTEM are enabled

When building with make 'ENABLE_JIT=1 ENABLE_SYSTEM=1', the following
error is observed:

src/emulate.c:45:13: error: ‘need_clear_block_map’ defined but not used [-Werror=unused-variable]
   45 | static bool need_clear_block_map = false;
      |             ^~~~~~~~~~~~~~~~~~~~

Fix this by guarding the 'need_clear_block_map' definition with the
RV32_HAS(SYSTEM) and !RV32_HAS(JIT) macros. This ensures the variable
is only defined when the macros indicate it is needed.

eleanorLYJ committed:
Fix undeclared identifier 'attr' error

Building with 'ENABLE_JIT=1, ENABLE_SYSTEM=1, and ENABLE_ELF_LOADER=0'
causes the following errors:
src/riscv.c:582:18: error: use of undeclared identifier 'attr'
  582 |     u8250_delete(attr->uart);
      |                  ^
src/riscv.c:583:17: error: use of undeclared identifier 'attr'
  583 |     plic_delete(attr->plic);
      |                 ^

These errors occur because attr is not defined under these conditions.
Fix this by adjusting the guard for attr to
(RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)), ensuring it is defined as
needed.

visitorckw commented:
> Hmmm, it seems CI fails due to insufficient memory. Any ideas on how we can reduce the memory usage?

Sorry for the misleading description. The CI failure is due to the addition of more tests, causing excessive http requests to github.com during the fetch-checksum, which results in a 403 too many requests error. Here’s the error log:

```
--2025-01-10 17:19:46--  https://api.github.com/repos/sysprog21/rv32emu-prebuilt/releases
Resolving api.github.com (api.github.com)... 140.82.114.6
Connecting to api.github.com (api.github.com)|140.82.114.6|:443... connected.
HTTP request sent, awaiting response... 403 rate limit exceeded
2025-01-10 17:19:46 ERROR 403: rate limit exceeded.
```

eleanorLYJ commented:
Thank you for your help. I will fix it.

eleanorLYJ commented:
> Sorry for the misleading description. The CI failure is due to the addition of more tests, causing excessive http requests to github.com during the fetch-checksum, which results in a 403 too many requests error. Here’s the error log:
> 
> ```
> --2025-01-10 17:19:46--  https://api.github.com/repos/sysprog21/rv32emu-prebuilt/releases
> Resolving api.github.com (api.github.com)... 140.82.114.6
> Connecting to api.github.com (api.github.com)|140.82.114.6|:443... connected.
> HTTP request sent, awaiting response... 403 rate limit exceeded
> 2025-01-10 17:19:46 ERROR 403: rate limit exceeded.
> ```

I noticed that we make frequent requests to fetch the checksum file and store it in $(BIN_DIR)/sha1sum-linux-image.

To reduce the number of requests, we could modify the code to check if the checksum file already exists and is less than a day old. If it’s up-to-date, we can skip fetching it again. This way, we minimize unnecessary requests.

Any thoughts?

jserv commented:
> I noticed that we make frequent requests to fetch the checksum file and store it in $(BIN_DIR)/sha1sum-linux-image.
> 
> To reduce the number of requests, we could modify the code to check if the checksum file already exists and is less than a day old. If it’s up-to-date, we can skip fetching it again. This way, we minimize unnecessary requests.

Yes, create an issue for tracking.
Cc. @vacantron

eleanorLYJ committed:
CI: Add test case for disabling Zifencei

Add a new test case to the CI pipeline when the Zifencei is disabled.
This ensures better coverage and prevents regressions for this specific
configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zicsr with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zicsr extension is disabled. This ensures better coverage and
prevents regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling Zifencei with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the Zifencei extension is disabled. This ensures better coverage and
prevents regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling MOP_FUSION with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the MOP_FUSION extension is disabled. This ensures better coverage and
prevents regressions for this specific configuration.

eleanorLYJ committed:
CI: Add test case for disabling BLOCK_CHAINING with JIT enabled

Add a new test case to the CI pipeline to verify JIT functionality when
the BLOCK_CHAINING extension is disabled. This ensures better coverage
and prevents regressions for this specific configuration.

ChinYikMing commented:
I'll help review the pull request this week. Thanks.

visitorckw reviewed (approved):
Except for the occasional rate limit exceed issue when fetching checksums in CI, everything else LGTM!

I'm surprised there are so many compilation issues. Thanks for your work!

ChinYikMing commented on src/emulate.c:
```diff
@@ -688,6 +708,7 @@ static void block_translate(riscv_t *rv, block_t *block)
         remove_next_nth_ir(rv, ir, block, count - 1);             \
     }
 
+#if RV32_HAS(MOP_FUSION)
```
Move this to line 686, as `COMBINE_MEM_OPS` macro is only utilized when `MOP_FUSION` is enabled and for readability, such that:
```c
#if RV32_HAS(MOP_FUSION)
#define COMBINE_MEM_OPS(RW)
...
```

ChinYikMing commented on src/emulate.c:
```diff
@@ -702,7 +723,9 @@ static inline void remove_next_nth_ir(const riscv_t *rv,
         block->ir_tail = ir;
     block->n_insn -= n;
 }
+#endif
```
Remove this since line 826 cover the condition.

eleanorLYJ replied:
Ack

ChinYikMing commented on src/emulate.c:
```diff
@@ -702,7 +723,9 @@ static inline void remove_next_nth_ir(const riscv_t *rv,
 }
+#endif
 
+#if RV32_HAS(MOP_FUSION)
```
Remove this since line 686 cover this.

eleanorLYJ replied:
Thank you for your comment.
I noticed that line 686 isn’t protected by the macro (if RV32_HAS(MOP_FUSION)). Perhaps you mean the code starting from line 711 is protected?

ChinYikMing replied:
> Thank you for your comment. I noticed that line 686 isn’t protected by the macro (if RV32_HAS(MOP_FUSION)). Perhaps you mean the code starting from line 711 is protected?

Check this:
```c
#if RV32_HAS(MOP_FUSION)
#define COMBINE_MEM_OPS(RW)
...
```

eleanorLYJ replied:
You're right. The macro should start covering from line 686 instead, as 'remove_next_nth_ir' is used inside the '#define COMBINE_MEM_OPS(RW)'.

ChinYikMing commented on src/emulate.c:
```diff
@@ -79,12 +81,14 @@ static void __trap_handler(riscv_t *rv);
     rv->csr_time[0] = ctr & 0xFFFFFFFF;
     rv->csr_time[1] = ctr >> 32;
 }
+#endif
```
Remove this since line 84 covers the condition.

eleanorLYJ replied:
Ack

ChinYikMing commented on src/emulate.c:
```diff
@@ -79,12 +81,14 @@ static void __trap_handler(riscv_t *rv);
 }
+#endif
 
 #if RV32_HAS(Zicsr)
```
Ditto.

eleanorLYJ replied:
Ack

ChinYikMing commented:
May I know if the remaining tests will still be executed in case any single test fails?

eleanorLYJ commented:
> May I know if the remaining tests will still be executed in case any single test fails?

Based on the current configuration, if any single test fails, the remaining tests will not be executed. Should we consider allowing the other tests to continue running even if one test fails?

ChinYikMing commented:
> > May I know if the remaining tests will still be executed in case any single test fails?
> 
> Based on the current configuration, if any single test fails, the remaining tests will not be executed. Should we consider allowing the other tests to continue running even if one test fails?

Yes, I think revealing all errors or warnings at once for each build flag is a better approach. Moreover, the validation process for both compilers only takes around 3 to 5 minutes based on previous actions log.

ChinYikMing commented:
In 'mk/toolchain.mk', we have `emcc` (Emscripten). @eleanorLYJ Can you also add the `emcc` to the CI? The reason is that 'deploy-wasm.yml' is only executed when specific files are changed or manually triggered. By detecting warnings or errors early, we can address them before the actual execution of 'deploy-wasm.yml'.

eleanorLYJ committed:
CI: Add 'if: always()' for steps

Ensure that subsequent steps are executed regardless of the success or
failure of previous steps while providing more explicit control over step
execution.

eleanorLYJ commented:
> In 'mk/toolchain.mk', we have `emcc` (Emscripten). @eleanorLYJ Can you also add the `emcc` to the CI? The reason is that 'deploy-wasm.yml' is only executed when specific files are changed or manually triggered. By detecting warnings or errors early, we can address them before the actual execution of 'deploy-wasm.yml'.

The current action we are using, `rlalik/setup-cpp-compiler`, does not support installing `emcc` at the moment. I plan to address this limitation in the next PR by implementing functionality to include `emcc` in the CI setup.

ChinYikMing commented:
> > In 'mk/toolchain.mk', we have `emcc` (Emscripten). @eleanorLYJ Can you also add the `emcc` to the CI? The reason is that 'deploy-wasm.yml' is only executed when specific files are changed or manually triggered. By detecting warnings or errors early, we can address them before the actual execution of 'deploy-wasm.yml'.
> 
> The current action we are using, `rlalik/setup-cpp-compiler`, does not support installing `emcc` at the moment. I plan to address this limitation in the next PR by implementing functionality to include `emcc` in the CI setup.

Got it. Looking forward to that!

ChinYikMing reviewed (approved):
LGTM.

jserv commented:
Thank @eleanorLYJ for contributing!
